### PR TITLE
Convert clock and time functions to pybind11

### PR DESCRIPTION
**Pull request by sloretz**
Part of #665

Converts time_point and clock functions to pybind11. Along the way I fixed a signed/unsigned bug where python to C conversion allowed time values up to 2\*\*64-1 to be set to be set, but the `rcl_time_point_t` struct only accepts up to 2\*\*63-1 because the type is signed. I also replaced the `ClockType` and `ClockChange` enums with pybind11 ones to make the conversion between the C and python layer easier.

* rclpy_create_time_point
* rclpy_time_point_get_nanoseconds
* rclpy_create_clock
* rclpy_clock_get_now
* rclpy_clock_get_ros_time_override_is_enabled
* rclpy_clock_set_ros_time_override_is_enabled
* rclpy_clock_set_ros_time_override
* rclpy_add_clock_callback, METH_VARARGS
* rclpy_remove_clock_callback, METH_VARARGS

### CODE REVIEW

File: rclpy/src/rclpy/_rclpy_pybind11.cpp [hunk 1]
```diff
@@ -14,15 +14,37 @@
...
+    "ROS_TIME_NO_CHANGE", RCL_ROS_TIME_NO_CHANGE,
+    "ROS time is active and will continue to be active")
+  .value(
+    "ROS_TIME_ACTIVATED", RCL_ROS_TIME_ACTIVATED,
+    "ROS time is being activated")
+  .value(
+    "ROS_TIME_DEACTIVATED", RCL_ROS_TIME_DEACTIVATED,
+    "ROS TIME is being deactivated, the clock will report system time after the jump")
+  .value(
+    "SYSTEM_TIME_NO_CHANGE ", RCL_SYSTEM_TIME_NO_CHANGE,
```
Comment (cottsay) [inline rclpy/src/rclpy/_rclpy_pybind11.cpp:45]: There is an excess space here:
```suggestion
    "SYSTEM_TIME_NO_CHANGE", RCL_SYSTEM_TIME_NO_CHANGE,
```

File: rclpy/src/rclpy/_rclpy_pybind11.cpp [hunk 2]
```diff
@@ -14,15 +14,37 @@
...
 #include "rclpy_common/exceptions.hpp"
 #include "service_info.hpp"
+#include "time_point.hpp"
 
 namespace py = pybind11;
 
 PYBIND11_MODULE(_rclpy_pybind11, m) {
   m.doc() = "ROS 2 Python client library.";
 
+  py::enum_<rcl_clock_type_t>(m, "ClockType")
```
Comment (cottsay) [inline rclpy/src/rclpy/_rclpy_pybind11.cpp:28]: It's nice to see these values backed by RCL values.

File: rclpy/src/rclpy/time_point.cpp [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Include pybind11 before rclpy_common/handle.h includes Python.h
+#include <pybind11/pybind11.h>
+
+#include <rcl/rcl.h>
+
+#include <cstring>
```
Comment (cottsay) [inline rclpy/src/rclpy/time_point.cpp:20]: What is this needed for?

File: rclpy/src/rclpy/clock.cpp [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+  } else {
+    ret = rcl_disable_ros_time_override(clock);
+  }
+  if (ret != RCL_RET_OK) {
+    throw RCLError("failed to set ROS time override");
+  }
+  if (PyErr_Occurred()) {
+    // Time jump callbacks raised
+    throw py::error_already_set();
+  }
```
Comment (cottsay) [inline rclpy/src/rclpy/clock.cpp:145]: Could another unrelated failure cause `PyErr_Occurred` to be true? Is this something we could tackle in the class rewrite?

File: rclpy/src/rclpy/clock.cpp [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+  if (!clock) {
+    throw py::error_already_set();
+  }
+
+  bool is_enabled;
+  rcl_ret_t ret = rcl_is_enabled_ros_time_override(clock, &is_enabled);
+  if (ret != RCL_RET_OK) {
+    throw RCLError("failed to get ROS time override status");
+  }
+
```
Comment (cottsay) [inline rclpy/src/rclpy/clock.cpp:120]: Should this function be checking `PyErr_Occurred` as well?

**Review by cottsay — APPROVED** [review]
Some comments, but no blockers. Looks good!

File: rclpy/src/rclpy/time_point.cpp (hunk 3 above)
Comment (sloretz) [reply rclpy/src/rclpy/time_point.cpp:20]: It's for [`std::strcmp`](https://github.com/ros2/rclpy/pull/699/files/e8fb88a5b29a07137badef80d28dfebf033e8cbf#diff-2e12c9701227e94447565daeac8ceef7b9e5997b09b09617e1d3c4bcb162c910R75)

File: rclpy/src/rclpy/clock.cpp (hunk 5 above)
Comment (sloretz) [reply rclpy/src/rclpy/clock.cpp:120]: I don't think so. Checking is only necessary when arbitrary python code can be run, like time jump callbacks. [`rcl_is_enabled_ros_time_override()` doesn't invoke arbitrary python code](https://github.com/ros2/rcl/blob/4740c82864518a331ae98799f25b2ba085b22473/rcl/src/rcl/time.c#L333).

File: rclpy/rclpy/time.py [hunk 6]
```diff
@@ -32,11 +32,11 @@ def __init__(self, *, seconds=0, nanoseconds=0, clock_type=ClockType.SYSTEM_TIME
             raise ValueError('Nanoseconds value must not be negative')
         total_nanoseconds = int(seconds * CONVERSION_CONSTANT)
         total_nanoseconds += int(nanoseconds)
-        try:
-            self._time_handle = _rclpy.rclpy_create_time_point(total_nanoseconds, clock_type)
-        except OverflowError as e:
+        if total_nanoseconds >= 2**63:
+            # pybind11 would raise TypeError, but we want OverflowError
```
[the remark this answers is not in the document]
Comment (sloretz) [reply rclpy/rclpy/time.py:36]: As in, would there be an advantage in wrapping the `rclpy_create_time_point()` with a `try/except TypeError` and then moving the logic to re-raise OverflowError in there?  I can't think of one. Got an idea in mind?

[the remark this answers is not in the document]
Comment (cottsay) [reply rclpy/rclpy/time.py:36]: Nah, just curious.

File: rclpy/src/rclpy/_rclpy_pybind11.cpp (hunk 1 above)
Comment (sloretz) [reply rclpy/src/rclpy/_rclpy_pybind11.cpp]: Good catch! https://github.com/ros2/rclpy/pull/699/commits/2560af0471d8635c6620c95fe5e97735f4c7d731

**Comment by sloretz** [comment]
CI (build: `--packages-up-to rclpy` test: `--packages-select rclpy`)
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13836/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8695/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11544/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/13920/)

**Comment by sloretz** [comment]
PR job looks like a flaky test. Merging!

**Comment by sloretz** [comment]
@ros-pull-request-builder retest this please

**Pull request merged**